src/lib/shmem/src/raw_syscall.rs

```diff
@@ -0,0 +1,222 @@
+//! This module provides a simple interface to make system calls without relying on libc or nix,
+//! which are not no-std. The goal of this module is to be as no-std as possible to be compatible
+//! with Shadow's shim preload library.
+//!
+//! Public functions for system calls are named according to their standard names found in man(2).
+//! The public function accept have parameters with types that are natural for Rust. Most of these
+//! public functions are accompanied by a *_impl() function that has a signature which more closely
+//! matches the x86_64 ABI:
+//! <https://blog.rchapman.org/posts/Linux_System_Call_Table_for_x86_64/>
+
+use std::arch::asm;
+
+unsafe fn interpret_str_to_char_ptr(s: &str) -> *const char {
+    unsafe { core::mem::transmute::<*const u8, *const char>(s.as_bytes().as_ptr()) }
+}
+
+fn interpret_i32_as_u64(x: i32) -> u64 {
+    u64::from_ne_bytes((x as i64).to_ne_bytes())
+}
+
+fn interpret_u64_as_i32(x: u64) -> i32 {
+    let x_bytes = x.to_ne_bytes();
+    if cfg!(target_endian = "big") {
+        i32::from_ne_bytes(x_bytes[4..8].try_into().unwrap())
+    } else {
+        i32::from_ne_bytes(x_bytes[0..4].try_into().unwrap())
+    }
+}
+
+fn convert_i32_rv_to_rv_errno(rv: i32) -> Result<i32, i32> {
+    if rv < -1 {
+        Err(-1 * rv)
+    } else {
+        Ok(rv)
+    }
+}
+
+fn x86_64_syscall(
+    syscall_number: u64,
+    arg1: u64,
+    arg2: u64,
+    arg3: u64,
+    arg4: u64,
+    arg5: u64,
+    arg6: u64,
+) -> u64 {
+    let result: u64;
+
+    unsafe {
+        asm!(
+            "syscall",
+            inlateout("rax") syscall_number => result,
+            inlateout("rdi") arg1 => _,
+            inlateout("rsi") arg2 => _,
+            inlateout("rdx") arg3 => _,
+            inlateout("r10") arg4 => _,
+            inlateout("r8") arg5 => _,
+            inlateout("r9") arg6 => _,
+            lateout("rcx") _,
+            lateout("r11") _,
+        );
+    }
+
+    result
+}
+
+fn open_impl(filename: *const char, flags: i32, mode: u32) -> i32 {
+    const SYS_OPEN: u64 = 2;
+
+    let filename = unsafe { core::mem::transmute::<*const char, u64>(filename) };
+
+    interpret_u64_as_i32(x86_64_syscall(
+        SYS_OPEN,
+        filename,
+        interpret_i32_as_u64(flags),
+        mode as u64,
+        0,
+        0,
+        0,
+    ))
+}
+
+pub fn open(filename: &str, flags: i32, mode: u32) -> Result<i32, i32> {
```

Comment:
I was going to suggest  taking a `&Path` here but whoops of course that's `std` :). In this case it's probably better to take something like a `&[u8]` and let the caller worry about any necessary conversion and ensuring that it's null-terminated. (But it'd be worth having an assertion or check here to ensure it really is NULL-terminated)

Ultimately it'd be nice use [`bitflags`](https://crates.io/crates/bitflags) to generate types to use here, and maybe to avoid pulling in the `libc` crate by generating our own bindings from the kernel headers. I started to do the latter in the [crate I'm sketching out](https://github.com/shadow/shadow/pull/2930)